src/library/scala/collection/immutable/TreeSet.scala

```diff
@@ -119,11 +119,14 @@ final class TreeSet[A] private (tree: RB.Tree[A, Unit])(implicit val ordering: O
 
   def this()(implicit ordering: Ordering[A]) = this(null)(ordering)
 
-  private def newSet(t: RB.Tree[A, Unit]) = new TreeSet[A](t)
+  private def newSet(t: RB.Tree[A, Unit]) = {
+    if (t eq this.tree) this
+    else new TreeSet[A](t)
+  }
 
   /** A factory to create empty sets of the same type of keys.
    */
-  override def empty = TreeSet.empty
+  override def empty = newSet(null)
```

Comment:
Why would you want to create a new instance every time instead of reusing the `TreeSet.empty` singleton?

Reply:
TreeSet isn't a singleton, it has the implicit ordering parameter so creates a new instance
newSet will not create a new Set if `this` is already empty, and that is the best that we can do as I see it
` def empty[A](implicit ordering: Ordering[A]) = new TreeSet[A]`